gh-119933 : Improve ``SyntaxError`` message for invalid type parameters expressions

picnixz opened the pull request:
cc @JelleZijlstra 

I added a bunch of tests but feel free to tell me if there are more cases to check. It took me a bit of time to observe that entering the context was handled inside  `symtable_visit_type_param_bound_or_default` and not upon calling it.

Tell me if I need a What's New entry by the way or if the NEWS entry should be improved.

<!-- gh-issue-number: gh-119933 -->
* Issue: gh-119933
<!-- /gh-issue-number -->

picnixz committed:
fix ``SyntaxError`` for invalid type parameters expressions

picnixz committed:
blurb

picnixz committed:
improve doc & naming

picnixz committed:
visual improvements of some test

JelleZijlstra commented on Include/internal/pycore_symtable.h:
```diff
@@ -83,6 +107,7 @@ typedef struct _symtable_entry {
     PyObject *ste_directives;/* locations of global and nonlocal statements */
     PyObject *ste_mangled_names; /* set of names for which mangling should be applied */
     _Py_block_ty ste_type;
+    _extra_flags_t ste_extra_flags; /* extra temporary flags */
```
What do you think of adding a `char *` here with the "scope name" instead of this flag? I feel like that would simplify things since you don't need to keep track of flag values.

If we go with that, we can also get rid of the enum entries for type aliases and type params, and simply pass a different scope name string when creating the symbol table entry.

picnixz replied:
Mmh, if we are only using it for error messages, then I think it is a good idea. Should i actually remove everything in one PR or should I (for now) only focus on using the current interface + scope name without changing what you've written?

JelleZijlstra replied:
Yes, the distinction between TypeVarBoundBlock, TypeAliasBlock, and TypeParamBlock is purely for error messages; they work the same. We can make all the changes in this PR.

picnixz replied:
But what about the exported constants in the `symtable` module? and https://docs.python.org/3/library/symtable.html#symtable.SymbolTable.get_type? wouldn't it break some public API?

JelleZijlstra replied:
The C API is private and we can change it, but you're right about the symtable module. I suppose "TypeVar bound" there is also incorrect.

picnixz replied:
I feel that I should maybe change into "type parameter bound/constraint/default" (I'm leaving now, so I'll do the changes tomorrow, feel free to tell me where I misused some of the helper string functions btw).

JelleZijlstra commented on Python/symtable.c:
```diff
@@ -1339,6 +1340,8 @@ symtable_enter_block(struct symtable *st, identifier name, _Py_block_ty block,
     if (prev) {
         ste->ste_comp_iter_expr = prev->ste_comp_iter_expr;
+        /* inherit already set flags */
+        ste->ste_extra_flags = prev->ste_extra_flags;
```
They shouldn't be inherited. E.g., this is fine:

```
>>> def f[T: (lambda: (yield))](): pass
... 
```

picnixz replied:
Oh yes. Actually I found an issue with my construction so I'm fixing it now (I'll go with the `char*` alternative because it would simplify the flow).

picnixz committed:
simplify the flow

picnixz committed:
Merge branch 'main' into fix-119933

picnixz committed:
Merge branch 'main' into fix-119933

picnixz committed:
address review

- improve comments
- rename 'TypeVarBoundBlock' -> 'TypeVariableBlock'
- rename 'TypeParamBlock' -> 'TypeParametersBlock'
- rename 'ste_description' -> 'ste_context_info'
- only set 'ste_context_info' just before visiting the actual expression

picnixz committed:
address review

- rename '_symtable.TYPE_TYPE_VAR_BOUND' to '_symtable.TYPE_TYPE_VARIABLE'
- rename '_symtable.TYPE_TYPE_PARAM' to '_symtable.TYPE_TYPE_PARAMETERS'
- add string enumeration for symbol table type

picnixz committed:
use the same declaration order across files

JelleZijlstra commented on Include/internal/pycore_symtable.h:
```diff
@@ -15,11 +15,21 @@ typedef enum _block_type {
+    // The block to enter when processing a "generic" (PEP 695) object,
+    // e.g., "def foo[T](): pass" or "class A[T]: pass".
+    TypeParametersBlock,
+    // The block to enter when processing a single "type variable" in the
```
This doesn't tell the whole story as there may be more than one block per type variable (`def f[T: A = B](): pass`).
```suggestion
    // The block to enter when processing the bounds, constraints, or
    // default value of a single "type variable" in the
```

picnixz replied:
Thank you, I wasn't precise enough. By "single", I meant  that whenever you process a type variable bound, constraints or default, then you enter that block. I didn't want to say "single" in the sense "T: A = B" is a single TypeVariableBlock but rather processing "A" and "B" would create two different TypeVariableBlock.

JelleZijlstra commented on Lib/symtable.py:
```diff
@@ -66,23 +77,23 @@ def __repr__(self):
         """
         if self._table.type == _symtable.TYPE_MODULE:
-            return "module"
+            return SymbolTableType.MODULE.value
```
Why not this?
```suggestion
            return SymbolTableType.MODULE
```
It still returns a string equal to `"module"`, so this should be fine.

picnixz replied:
Good catch! my bad.

JelleZijlstra commented on Lib/symtable.py:
```diff
@@ -66,23 +77,23 @@ def __repr__(self):
 
-        The values returned are 'class', 'module', 'function',
-        'annotation', 'TypeVar bound', 'type alias', and 'type parameter'.
+        The values returned are one of the values in
```
This should also be mentioned in the documentation (`symtable.rst`).

picnixz replied:
Oh right, I forgot to update it. Thanks!

JelleZijlstra commented on Misc/NEWS.d/next/Library/2024-06-03-13-48-44.gh-issue-119933.Kc0HG5.rst:
```diff
@@ -0,0 +1 @@
+Improve :exc:`SyntaxError` messages for :pep:`695` and :pep:`696`.
```
Should also mention the changes to the symtable module, and reference the features by name instead of the PEPs.

JelleZijlstra commented on Python/symtable.c:
```diff
@@ -2298,16 +2300,35 @@ symtable_visit_expr(struct symtable *st, expr_ty e)
         }
+
+        PyObject *scope_info;
+        scope_info = PyUnicode_FromFormat("%s %s", tp_kind_name,
```
Would be simpler to pass the whole `ste_scope_info` string into this function so you don't have to create it dynamically.

The current code appears to be leaking the value of `scope_info`. (Tip: run `./python -m test -R 3:3 test_type_params`, or some other test, to check for refleaks.)

picnixz replied:
Oh yes thank you. I actually struggled a bit to understand how to create a string and whether there is a mechanism for deallocating them correctly (maybe you could point some section I could read about that?).

By the way, on the main branch I have:

```text
Using random seed: 831335327
0:00:00 load avg: 2.20 Run 1 test sequentially in a single process
0:00:00 load avg: 2.20 [1/1] test_type_params
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
XX. ...
```

Is it fine that there are leaks? (the test does not fail but I would like to know why and whether I should bother).

JelleZijlstra replied:
Leaks on the first two ("warmup") runs are usually expected; various caches are being filled up. We care that there are no leaks on the last few runs.

JelleZijlstra replied:
As for deallocating them, you're creating a PyObject, so just `Py_DECREF` drops the reference and deallocates the string you've created.

picnixz replied:
Ah yes. Actually, I know what I did wrong. I think I read "The caller is **NOT** responsible for decref'ing the returned objects." and I was like "Oh great!"...

EDIT: In the end, I took your approach of creating the string before: easier and more efficient.

picnixz committed:
Use enumeration members instead of `.value`.

picnixz committed:
export `SymbolTableType` enumeration

picnixz committed:
update NEWS

picnixz committed:
simplify `ste_scope_info` creation

picnixz committed:
improve documentation for `TypeVariableBlock`

picnixz committed:
update documentation

picnixz committed:
fixup

picnixz committed:
Merge branch 'main' into fix-119933

picnixz committed:
blurb

picnixz committed:
fixup

picnixz committed:
Merge branch 'main' into fix-119933

picnixz committed:
Merge branch 'main' into fix-119933

JelleZijlstra commented on Lib/symtable.py:
```diff
@@ -69,23 +80,23 @@ def __repr__(self):
 
-        The values returned are 'class', 'module', 'function',
-        'annotation', 'TypeVar bound', 'type alias', and 'type parameter'.
+        The values returned are one of the values in
```
```suggestion
        The value returned is one of the members of
```

JelleZijlstra commented on Misc/NEWS.d/next/Library/2024-06-03-13-48-44.gh-issue-119933.Kc0HG5.rst:
```diff
@@ -0,0 +1,4 @@
+Improve :exc:`SyntaxError` messages for invalid expressions in a type
```
This should be in the Core and Builtins section, not Library.

JelleZijlstra commented on Lib/test/test_symtable.py:
```diff
@@ -87,13 +88,14 @@ def test_type(self):
         self.assertEqual(self.foo.get_type(), "function")
         self.assertEqual(self.Alias.get_type(), "type alias")
-        self.assertEqual(self.GenericAlias.get_type(), "type parameter")
+        self.assertEqual(self.GenericAlias.get_type(), "type parameters")
```
This is arguably a compatibility break with 3.12. Unfortunately, that's hard to entirely avoid. Perhaps we should make the documentation say that the exact values of the strings are not guaranteed and may change.

picnixz replied:
Should I do it for the backports only? or should I also mention it in 3.13 and 3.14?

picnixz replied:
Here is what I suggest:

- 3.12: use enumeration members with *old* values (to avoid breakage) but still use enumeration values. use `.. versionchanged:: 3.12` for this specific version.

- 3.13, 3.14: use enumeration members with *new* values (both versions are not stable so we should be able to use whatever values we want). Use `.. versionchanged:: 3.13` instead of `.. versionchanged:: 3.13` for both verions.

More precisely, here is what the RST would be:

**For 3.12**

```diff
+   .. attribute:: TYPE_ALIAS
+      :value: "type alias"
(...)
+   .. attribute:: TYPE_PARAMETERS
+      :value: "type parameter"
(...)
+
+   .. attribute:: TYPE_VARIABLE
+      :value: "TypeVar bound"
(...)
-      Return the type of the symbol table.  Possible values are ``'class'``,
-      ``'module'``, ``'function'``, ``'annotation'``, ``'TypeVar bound'``,
-      ``'type alias'``, and ``'type parameter'``. The latter four refer to
-      different flavors of :ref:`annotation scopes <annotation-scopes>`.
+      Return the type of the symbol table.  Possible values are members
+      of the :class:`SymbolTableType` enumeration.

       .. versionchanged:: 3.12
          Added ``'annotation'``,  ``'TypeVar bound'``, ``'type alias'``,
          and ``'type parameter'`` as possible return values.

+      .. versionchanged:: 3.12
+         Return values are members of the :class:`SymbolTableType` enumeration.
+
+         The exact values of the returned string may change in the future,
+         and thus, it is recommended to use :class:`SymbolTableType` members
+         instead of hard-coded strings.
```

**For 3.13 and 3.14**

```diff
+   .. attribute:: TYPE_ALIAS
+      :value: "type alias"
(...)
+   .. attribute:: TYPE_PARAMETERS
+      :value: "type parameters"
(...)
+
+   .. attribute:: TYPE_VARIABLE
+      :value: "type variable"
(...)
-      Return the type of the symbol table.  Possible values are ``'class'``,
-      ``'module'``, ``'function'``, ``'annotation'``, ``'TypeVar bound'``,
-      ``'type alias'``, and ``'type parameter'``. The latter four refer to
-      different flavors of :ref:`annotation scopes <annotation-scopes>`.
+      Return the type of the symbol table.  Possible values are members
+      of the :class:`SymbolTableType` enumeration.

       .. versionchanged:: 3.12
          Added ``'annotation'``,  ``'TypeVar bound'``, ``'type alias'``,
          and ``'type parameter'`` as possible return values.

+      .. versionchanged:: 3.13
+         Return values are members of the :class:`SymbolTableType` enumeration.
+
+         The exact values of the returned string may change in the future,
+         and thus, it is recommended to use :class:`SymbolTableType` members
+         instead of hard-coded strings.
```

And in the tests, we would check that separately depending on the branch we are working on.

JelleZijlstra replied:
I'd prefer to leave 3.12 unchanged; adding an enum seems too much in a bugfix release, and the bug this issue fixes isn't as bad there.

picnixz committed:
Move SyntaxError's NEWS into "Core and Builtins"

picnixz committed:
improve documentation wording

carljm reviewed (commented):
General direction looks fine to me. I won't have time soon to do a detailed review but if @JelleZijlstra is happy with it, that's good enough for me.

JelleZijlstra commented on Doc/library/symtable.rst:
```diff
@@ -31,21 +31,74 @@ Generating Symbol Tables
+      a TypeVar, a TypeVarTuple or a ParamSpec object (the latter two do
+      not support a bound or a constraint tuple).
+
+   .. versionadded:: 3.14
```
```suggestion
   .. versionadded:: 3.13
```

JelleZijlstra commented on Python/symtable.c:
```diff
@@ -2372,36 +2382,46 @@ symtable_visit_type_param(struct symtable *st, type_param_ty tp)
+        ste_scope_info = "a TypeVar default";
         if (!symtable_visit_type_param_bound_or_default(st, tp->v.TypeVar.default_value, tp->v.TypeVar.name,
-                                                        (void *)((uintptr_t)tp + 1))) {
+                                                        (void *)((uintptr_t)tp + 1), ste_scope_info)) {
```
```suggestion
                                                        (void *)((uintptr_t)tp + 1), "a TypeVar default")) {
```

picnixz replied:
Oups I did not catch that one. Thank you! I did not add such test, so I'll add it.

JelleZijlstra commented on Python/symtable.c:
```diff
@@ -2372,36 +2382,46 @@ symtable_visit_type_param(struct symtable *st, type_param_ty tp)
             VISIT_QUIT(st, 0);
         }
+
+        ste_scope_info = "a TypeVar default";
```
```suggestion
```

picnixz committed:
Update Python/symtable.c

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

picnixz committed:
Update Doc/library/symtable.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

picnixz committed:
address review
